File: node_receiver/src/jsonl.js
import { mkdir, appendFile } from "node:fs/promises";
import { join } from "node:path";

function utcNowIso() {
  return new Date().toISOString();
}

function dateSuffixUTC() {
  const now = new Date();
  const year = now.getUTCFullYear();
  const month = String(now.getUTCMonth() + 1).padStart(2, "0");
  const day = String(now.getUTCDate()).padStart(2, "0");
  return `${year}${month}${day}`;
}

export class JsonlWriter {
  constructor(logDir) {
    this.logDir = logDir;
    this.ready = mkdir(logDir, { recursive: true });
  }

  utcNowIso() {
    return utcNowIso();
  }

  async writeRaw(record) {
    await this.#write("raw", record);
  }

  async writeDecoded(record) {
    await this.#write("decoded", record);
  }

  async writeError(record) {
    await this.#write("errors", record);
  }

  async #write(stream, record) {
    await this.ready;
    const path = join(this.logDir, `${stream}-${dateSuffixUTC()}.jsonl`);
    await appendFile(path, `${JSON.stringify(record)}\n`, "utf8");
  }
}

File: python_receiver/src/rtu_receiver/jsonl.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Dict


class JsonlWriter:
    def __init__(self, log_dir: Path) -> None:
        self.log_dir = log_dir
        self.log_dir.mkdir(parents=True, exist_ok=True)

    @staticmethod
    def utc_now_iso() -> str:
        return datetime.now(tz=timezone.utc).isoformat()

    def write_raw(self, record: Dict[str, Any]) -> None:
        self._write("raw", record)

    def write_decoded(self, record: Dict[str, Any]) -> None:
        self._write("decoded", record)

    def write_error(self, record: Dict[str, Any]) -> None:
        self._write("errors", record)

    def _write(self, stream: str, record: Dict[str, Any]) -> None:
        date_suffix = datetime.now(tz=timezone.utc).strftime("%Y%m%d")
        path = self.log_dir / f"{stream}-{date_suffix}.jsonl"
        with path.open("a", encoding="utf-8") as fp:
            fp.write(json.dumps(record, ensure_ascii=False, separators=(",", ":")) + "\n")
